.github/workflows/src/summarize-checks/labelling.js

```diff
@@ -23,6 +72,220 @@ import {
+ */
+/**
+ * @typedef {Object} BreakingChangesCheckConfig
+ * @property {ReviewRequiredLabel} reviewRequiredLabel
+ * @property {ReviewApprovalPrefixLabel} approvalPrefixLabel
+ * @property {(ValidVersioningApproval | ValidBreakingChangeApproval)[]} approvalLabels
+ * @property {string} [deprecatedReviewRequiredLabel]
+ */
+
+/**
+ * @typedef {"SwaggerFile" | "TypeSpecFile" | "ExampleFile" | "ReadmeFile"} FileTypes
+ */
+
+/**
+ * @typedef {"Addition" | "Deletion" | "Update"} ChangeTypes
+ */
+
+/**
+ * @typedef {Object} PRChange
+ * @property {FileTypes} fileType
+ * @property {ChangeTypes} changeType
+ * @property {string} filePath
+ * @property {any} [additionalInfo]
+ */
+
+/**
+ * @typedef {Object} ChangeHandler
+ * @property {function(PRChange): void} [SwaggerFile]
+ * @property {function(PRChange): void} [TypeSpecFile]
+ * @property {function(PRChange): void} [ExampleFile]
+ * @property {function(PRChange): void} [ReadmeFile]
+ */
+
+/**
+ * @typedef {"resource-manager" | "data-plane"} PRType
+ */
+
+/**
+ * Represents a GitHub label.
+ * Currently used in the context of processing
+ * labels related to the review workflow of PRs submitted to
+ * Azure/azure-rest-api-specs and Azure/azure-rest-api-specs-pr repositories.
+ * This processing happens in the prSummary.ts / summary() function.
+ *
+ * See also: https://aka.ms/SpecPRReviewARMInvariants
+ */
+// todo: inject `core` for access to logging
+export class Label {
+  /**
+   * @param {string} name
+   * @param {Set<string>} [presentLabels]
+   */
+  constructor(name, presentLabels) {
+    /** @type {string} */
+    this.name = name;
+
+    /**
+     * Is the label currently present on the pull request?
+     * This is determined at the time of construction of this object.
+     * @type {boolean | undefined}
+     */
+    this.present = presentLabels?.has(this.name) ?? undefined;
+
+    /**
+     * Should this label be present on the pull request?
+     * Must be defined before applyStateChange is called.
+     * Not set at the construction time to facilitate determining desired presence
+     * of multiple labels in single code block, without intermixing it with
+     * label construction logic.
+     * @type {boolean | undefined}
+     */
+    this.shouldBePresent = undefined;
+  }
+
+  /**
+   * If the label should be added, add its name to labelsToAdd.
+   * If the label should be removed, add its name to labelsToRemove.
+   * Otherwise, do nothing.
+   *
+   * Precondition: this.shouldBePresent has been defined.
+   * @param {Set<string>} labelsToAdd
+   * @param {Set<string>} labelsToRemove
+   */
+  applyStateChange(labelsToAdd, labelsToRemove) {
+    if (this.shouldBePresent === undefined) {
+      console.warn(
+        "ASSERTION VIOLATION! " +
+          `Cannot applyStateChange for label '${this.name}' ` +
+          "as its desired presence hasn't been defined. Returning early."
+      );
+      return;
```

Comment:
We should probably throw instead of returning.  It's easier to find bugs if workflows crash on error.

Reply:
Done.